src/app/games/hive/HivePieceBehavior.ts

```diff
@@ -0,0 +1,299 @@
+        if (HexagonalUtils.areNeighbors(move.coord, move.end) === false) {
+            return MGPFallible.failure(HiveFailure.BEETLE_CAN_ONLY_MOVE_TO_DIRECT_NEIGHBORS());
+        }
+        return MGPFallible.success(undefined);
+    }
+
+    public getPossibleMoves(coord: Coord, state: HiveState): HiveMoveCoordToCoord[] {
+        const moves: HiveMoveCoordToCoord[] = [];
+        for (const neighbor of HexagonalUtils.getNeighbors(coord)) {
+            moves.push(HiveMoveCoordToCoord.from(coord, neighbor).get());
+        }
+        return moves;
+    }
+}
+
+export class HivePieceBehaviorGrasshopper extends HivePieceBehavior {
+
+    private static INSTANCE: MGPOptional<HivePieceBehaviorGrasshopper> = MGPOptional.empty();
+
+    public static get(): HivePieceBehaviorGrasshopper {
+        if (this.INSTANCE.isAbsent()) {
+            this.INSTANCE = MGPOptional.of(new HivePieceBehaviorGrasshopper());
+        }
+        return this.INSTANCE.get();
+    }
+
+    public moveLegality(move: HiveMoveCoordToCoord, state: HiveState): MGPFallible<void> {
+
+        const direction: MGPFallible<HexaDirection> = HexaDirection.factory.fromMove(move.coord, move.end);
+        if (direction.isFailure()) {
+            return MGPFallible.failure(HiveFailure.GRASSHOPPER_MUST_MOVE_IN_STRAIGHT_LINE());
+        }
+        const jumpedCoords: Coord[] = move.coord.getCoordsToward(move.end);
+        if (jumpedCoords.length === 0) {
+            return MGPFallible.failure(HiveFailure.GRASSHOPPER_MUST_JUMP_OVER_PIECES());
+        }
+        for (const coord of jumpedCoords) {
+            if (state.getAt(coord).isEmpty()) {
+                return MGPFallible.failure(HiveFailure.GRASSHOPPER_MUST_JUMP_OVER_PIECES());
+            }
+        }
+        return this.checkEmptyDestination(move, state);
+    }
+    public getPossibleMoves(coord: Coord, state: HiveState): HiveMoveCoordToCoord[] {
+        const moves: HiveMoveCoordToCoord[] = [];
+        for (const direction of HexaDirection.factory.all) {
+            const neighbor: Coord = coord.getNext(direction);
+            if (state.getAt(neighbor).isOccupied()) {
+                // We can jump in that direction
+                let end: Coord = neighbor;
+                while (state.getAt(end).isOccupied()) {
+                    end = end.getNext(direction);
+                }
+                moves.push(HiveMoveCoordToCoord.from(coord, end).get());
+            }
+        }
+        return moves;
+    }
+}
+
+export class HivePieceBehaviorSpider extends HivePieceBehavior {
+
+    private static INSTANCE: MGPOptional<HivePieceBehaviorSpider> = MGPOptional.empty();
+
+    public static get(): HivePieceBehaviorSpider {
+        if (this.INSTANCE.isAbsent()) {
+            this.INSTANCE = MGPOptional.of(new HivePieceBehaviorSpider());
+        }
+        return this.INSTANCE.get();
+    }
+
+    public prefixLegality(coords: Coord[], state: HiveState): MGPFallible<void> {
```

Comment:
mmh, not the best name is it ?
Aussi, il est où le check que ça devrait faire une liste de 3 précisément ?

Reply:
`prefixLegality` peut avoir une liste plus petite que 3
Il est appelé avec les coord du spiderMove (exactement 3, enforcé par les types) quand on check la légalité d'un mouvement
Il est appelé par le composant avec moins de 3 mouvements pour vérifier la légalité du préfixe du mouvement (d'où le nom)